httpcli/http.go

```diff
@@ -79,8 +98,11 @@ func (c *Client) Do(m encoding.Marshaler) (encoding.Decoder, io.Closer, error) {
 	req.Header.Set("Content-Type", c.codec.MediaTypes[0])
 	req.Header.Set("Accept", c.codec.MediaTypes[1])
 
-	res, err := c.cli.Do(req)
+	res, err := c.do(req)
 	if err != nil {
+		if res != nil && res.Body != nil {
```

Comment:
I don't get this. When there is an error, res should be always nil. This invariant must be part of the API.